cmd/appender/main.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"log"
+	"net/http"
+
+	"github.com/google/go-containerregistry/authn"
+	"github.com/google/go-containerregistry/name"
+	"github.com/google/go-containerregistry/v1"
+	"github.com/google/go-containerregistry/v1/remote"
+	"github.com/google/go-containerregistry/v1/tarball"
+	"github.com/google/go-containerregistry/v1/writable"
+)
+
+var (
+	inputFile  = flag.String("tarball", "", "the tarball to append")
+	srcImage   = flag.String("src-image", "", "The tag of the remote image to append to")
+	dstImage   = flag.String("dst-image", "", "The tag of the resulting remote image")
+	outputFile = flag.String("output-tarball", "", "The output file of the resulting image")
+)
+
+func main() {
+	flag.Parse()
+
+	validateInput()
+
+	src, err := name.NewTag(*srcImage, name.WeakValidation)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	srcAuth, err := authn.DefaultKeychain.Resolve(src.Registry)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	dst, err := name.NewTag(*dstImage, name.WeakValidation)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	i, err := remote.Image(src, srcAuth, http.DefaultTransport)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	w, err := writable.NewImage(i)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	if err := w.AppendLayer(*inputFile); err != nil {
+		log.Fatalln(err)
+	}
+
+	if *outputFile != "" {
+		writeTarball(dst, w)
+		return
+	}
+
+	writeRemote(dst, w)
+}
+
+func writeRemote(tag name.Tag, i v1.Image) {
+	dstAuth, err := authn.DefaultKeychain.Resolve(tag.Registry)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	if err := remote.Write(tag, i, dstAuth, http.DefaultTransport, remote.WriteOptions{}); err != nil {
```

Comment:
We certainly want to pass `MountPaths`, see: https://github.com/google/go-containerregistry/blob/master/v1/remote/write.go#L35

Reply:
tl;dr if you do this, the registry (when within the same domain) can fast-path blobs, and keep you from ever downloading them.

Reply:
I've seen mounting work on DockerHub _without_ specifying anything in the `remote.WriteOptions`

Since as part of initiating the upload of a blob includes the `mount` query param

Reply:
I wonder if the `from` query param is not actually needed for cross repo mounting. Does anyone (@jonjohnsonjr  @ImJasonH) know if gcr.io follow the same behaviour?

Reply:
Hi, I implemented it for GCR. If the blob is already in the destination repo, `mount` is sufficient, since we can just check the digest. If it's not, you need `from` to tell us where to look to find it.

So first push would be improved a lot if you include `from`, but `mount` is enough for subsequent pushes.

Reply:
Awesome - thanks for the input